src/main/scala/temple/DSL/semantics/Validator.scala

```diff
@@ -0,0 +1,96 @@
+package temple.DSL.semantics
+
+import temple.ast.AttributeType._
+import temple.ast.{Metadata, _}
+
+private class Validator private (templefile: Templefile) {
+
+  private lazy val allStructs: Set[String] =
+    templefile.services.flatMap { case serviceName -> service => Iterator(serviceName) ++ service.structs.keys }.toSet
+
+  private lazy val allServices: Set[String] = templefile.services.keys.toSet
+
+  private def validateService(service: ServiceBlock)(context: SemanticContext): Unit = {
+    service.structs.foreachEntry(context(validateStruct))
+    service.attributes.foreachEntry(context(validateAttribute))
+    service.metadata.foreach(validateMetadata(_, service.attributes)(context))
+  }
+
+  private def validateStruct(struct: StructBlock)(context: SemanticContext): Unit = {
+    struct.attributes.foreachEntry(context(validateAttribute))
+    struct.metadata.foreach(validateMetadata(_, struct.attributes)(context))
+  }
+
+  private def validateMetadata(metadata: Metadata, attributes: Map[String, Attribute] = Map())(
+    context: SemanticContext,
+  ): Unit =
+    metadata match {
+      case _: Metadata.TargetLanguage    => // Nothing to validate yet
+      case Metadata.TargetAuth(_)        => context.fail(s"TODO: figure out how this auth works")
+      case _: Metadata.ServiceLanguage   => // Nothing to validate yet
+      case _: Metadata.Database          => // Nothing to validate yet
+      case _: Metadata.ServiceAuth       => // Nothing to validate yet
+      case _: Metadata.Readable          => // Nothing to validate yet
+      case _: Metadata.Writable          => // Nothing to validate yet
+      case Metadata.Omit(_)              => // Nothing to validate yet
+      case Metadata.ServiceEnumerable(_) => // Nothing to validate yet
+      case _: Metadata.Provider          => // Nothing to validate yet
```

Comment:
When would there be validation done here?

Reply:
Good question, do any of these need validation?

Reply:
Can't think of any - except for the one you've written below. I'm happy to leave this like this though, means when we think of things to add it's easy to do so